OrbitProducer/LockFreeBufferCaptureEventProducerTest.cpp

```diff
@@ -0,0 +1,345 @@
+// Copyright (c) 2020 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include <gmock/gmock.h>
+#include <gtest/gtest.h>
+
+#include "FakeProducerSideService.h"
+#include "OrbitProducer/LockFreeBufferCaptureEventProducer.h"
+#include "absl/strings/str_format.h"
+#include "grpcpp/grpcpp.h"
+#include "producer_side_services.grpc.pb.h"
+
+namespace orbit_producer {
+
+namespace {
+
+class LockFreeBufferCaptureEventProducerImpl
+    : public LockFreeBufferCaptureEventProducer<std::string> {
+ protected:
+  orbit_grpc_protos::CaptureEvent TranslateIntermediateEvent(
+      std::string&& /*intermediate_event*/) override {
+    return orbit_grpc_protos::CaptureEvent{};
+  }
+};
+
+class LockFreeBufferCaptureEventProducerTest : public ::testing::Test {
+ protected:
+  void SetUp() override {
+    fake_service.emplace();
+
+    grpc::ServerBuilder builder;
+    builder.RegisterService(&fake_service.value());
+    fake_server = builder.BuildAndStart();
+    ASSERT_NE(fake_server, nullptr);
+
+    std::shared_ptr<grpc::Channel> channel =
+        fake_server->InProcessChannel(grpc::ChannelArguments{});
+
+    buffer_producer.emplace();
+    buffer_producer->BuildAndStart(channel);
+
+    // Leave some time for the ReceiveCommandsAndSendEvents RPC to actually happen.
+    std::this_thread::sleep_for(std::chrono::milliseconds(50));
+  }
+
+  void TearDown() override {
+    // Leave some time for all pending communication to finish.
+    std::this_thread::sleep_for(std::chrono::milliseconds(50));
+
+    buffer_producer->ShutdownAndWait();
+    buffer_producer.reset();
+  }
+
+  void StopService() {
+    fake_service->FinishAndDisallowRpc();
+    fake_server->Shutdown();
+    fake_server->Wait();
+
+    fake_service.reset();
+    fake_server.reset();
+  }
+
+  std::optional<FakeProducerSideService> fake_service;
```

Comment:
Member vars should end in `_`.

Reply:
As this is a test fixture, this was on purpose to avoid those underscores everywhere in the tests. But in all the examples I found of fixtures they use the trailing underscore so I adapted.